### PR TITLE
new static files service for Web Servers supported by latest hendrix version

### PR DESCRIPTION
**Pull request by vepkenez**

### CODE REVIEW

File: tests/acceptance/network/test_static_files.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            response = requests.get(
+                "https://{}/statics/test-never-make-a-file-with-this-name.js".format(node.rest_url()),
+                verify=cert_file
+            )
+            assert response.status_code == 200
+            assert "I am Javascript" in response.text
+            assert response.headers['Content-Type'] == 'application/javascript'
+            return node
+
+        def check_static_file_not_there(node, cert_file):
```
Comment (derekpierre) [inline tests/acceptance/network/test_static_files.py:54]: it's not clear to me that this method is being used...

Comment (vepkenez) [reply tests/acceptance/network/test_static_files.py:54]: Ha, you're right!

File: tests/acceptance/network/test_static_files.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            return node
+
+        try:
+            with open("test-cert", "wb") as f:
+                f.write(cert_bytes)
+            os.makedirs(os.path.join(STATICS_DIR), exist_ok=True)
+            with open(os.path.join(STATICS_DIR, 'test-never-make-a-file-with-this-name.js'), 'w+') as fout:
+                fout.write("console.log('I am Javascript')\n")
+                fout.close()
+            yield threads.deferToThread(check_static_service, node, "test-cert")
```
Comment (vepkenez) [inline tests/acceptance/network/test_static_files.py:70]:
```suggestion
            yield threads.deferToThread(check_static_service, node, "test-cert")
            yield threads.deferToThread(check_static_file_not_there, node, "test-cert")
```

File: tests/acceptance/network/test_static_files.py (hunk 1 above)
Comment (vepkenez) [reply tests/acceptance/network/test_static_files.py:54]: fixed!

**Comment by derekpierre** [comment]
@vepkenez Two tests are failing:
1. `deployers` I think @vzotova already fixed it - you just need to rebase over main
2. `interfaces` - I guess you relocked deps in order to get the latest version of hendrix? It is failing because you are running into the same issue @KPrasch and I had in #2220 regarding py-solc-x; see https://github.com/nucypher/nucypher/pull/2220#issuecomment-686531961

**Review by KPrasch — COMMENTED** [review]
@vepkenez - Is there a new dependency or version update added in this PR?

**Comment by vepkenez** [comment]
> @vepkenez - Is there a new dependency or version update added in this PR?

hendrix==3.4.0

**Comment by vepkenez** [comment]
Looks like I forgot to update the Pipfile in this build.  Doing that.

**Comment by KPrasch** [comment]
Conflicts caused by PR #2213

File: dev-requirements.txt [hunk 3]
```diff
@@ -1,50 +1,139 @@
...
 chardet==3.0.4
+click==7.1.2
+coincurve==13.0.0
+colorama==0.4.3
+constant-sorrow==0.1.0a9
+constantly==15.1.0
 coverage==5.2.1
+cryptography==3.1
+cytoolz==0.10.1; implementation_name == 'cpython'
+dateparser==0.7.6; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4'
```
Comment (derekpierre) [inline dev-requirements.txt:26]: it's strange that this extra `python_version` text is being added ...

Comment (KPrasch) [reply dev-requirements.txt:26]: Yes, I agree - was this build from OSX, or docker?

Comment (vepkenez) [reply dev-requirements.txt:26]: I give up.
I did exactly what @KPrasch said to do on discord.
in docker:
```
root@5a974c44fed9:/code# rm Pipfile.lock 
root@5a974c44fed9:/code# rm requirements.txt 
root@5a974c44fed9:/code# rm dev-requirements.txt 
root@5a974c44fed9:/code# make lock
```

Comment (KPrasch) [reply dev-requirements.txt:26]: Ill give it a try - I've been having good success making builds.

Comment (vepkenez) [reply dev-requirements.txt:26]: Actually everything is working now... maybe this is just some new pip thing.

Comment (vepkenez) [reply dev-requirements.txt:26]: I just rebuilt the docker image so it's fresh, now I'm trying it again and seeing what happens

File: nucypher/network/resources.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+import os
+from hendrix.facilities.resources import MediaResource
+
+def get_static_resources():
+    resources = []
+    if os.getenv('NUCYPHER_STATIC_FILES_ROOT'):
```
Comment (KPrasch) [inline nucypher/network/resources.py:23]: Is this environment variable the only way to configure the static files path?

File: nucypher/characters/chaotic.py [hunk 5]
```diff
@@ -283,7 +284,7 @@ def start(self,
             raise RuntimeError("Felix is already running.")
 
         self.start_time = maya.now()
-        payload = {"wsgi": self.rest_app, "http_port": port}
+        payload = {"wsgi": self.rest_app, "http_port": port, "resources": get_static_resources()}
```
Comment (KPrasch) [inline nucypher/characters/chaotic.py:287]: Ah yes - this is refreshing.

**Review by KPrasch — APPROVED** [review]
Excellent work!

**Review by derekpierre — APPROVED** [review]
🎸 .

Just to clarify, the Pipfile is updated here, but hendrix is already locked to 3.4.0 in main hence no change to the *requirements.txt files - https://github.com/nucypher/nucypher/blob/main/requirements.txt#L35

**Comment by vepkenez** [comment]
It's sad that most of the conversation on this PR was about requirements issues unrelated to the single requirement change actually material to this changeset :|

File: nucypher/network/resources.py (hunk 4 above)
Comment (vepkenez) [reply nucypher/network/resources.py:23]: Yes, and by default no statics will be served.
I think that is the best/safest/most convenient way to approach this.

File: tests/acceptance/network/test_static_files.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                verify=cert_file
+            )
+            assert response.status_code == 404
+            return node
+
+        try:
+            with open("test-cert", "wb") as f:
+                f.write(cert_bytes)
+            os.makedirs(os.path.join(STATICS_DIR), exist_ok=True)
+            with open(os.path.join(STATICS_DIR, 'test-never-make-a-file-with-this-name.js'), 'w+') as fout:
```
Comment (KPrasch) [inline tests/acceptance/network/test_static_files.py:67]: now I'm gonna just because :stuck_out_tongue:

Comment (vepkenez) [reply tests/acceptance/network/test_static_files.py:67]: Baited!

**Comment by KPrasch** [comment]
Rebase over master should fix the failing CI tests.

**Review by cygnusv — APPROVED** [review]
LGTM!

**Pull request merged**